File: src/main/java/linq/collections/enumerator/ArrayEnumerator.java
package linq.collections.enumerator;

import linq.Enumerator;

public class ArrayEnumerator<T> implements Enumerator<T> {
    private final T[] array;
    private int index;

    public ArrayEnumerator(T[] array) {
        this.array = array;
        index = -1;
    }

    @Override
    public boolean moveNext() {
        var index = this.index + 1;
        var length = array.length;
        if (index >= length) {
            this.index = length;
            return false;
        }
        this.index = index;
        return true;
    }

    @Override
    public T current() {
        return array[index];
    }
}

File: src/main/kotlin/linq/collections/enumerator/ArrayEnumerator.kt
package linq.collections.enumerator;

import linq.Enumerator;

class ArrayEnumerator<T>(private val array: Array<out T>) : Enumerator<T> {
    var index: Int = -1;

    override val current: T get() = array[index]

    override fun moveNext(): Boolean {
        val index = this.index + 1;
        val length = array.size;
        if (index >= length) {
            this.index = length;
            return false;
        }
        this.index = index;
        return true;
    }
}
